controllers/buildpipeline/buildpipeline_adapter.go

```diff
@@ -129,6 +133,29 @@ func (a *Adapter) EnsureSnapshotExists() (controller.OperationResult, error) {
 		return controller.RequeueWithError(err)
 	}
 
+	return a.removeFinalizerAndContinueProcessing()
+}
+
+func (a *Adapter) EnsurePipelineIsFinalized() (controller.OperationResult, error) {
+	if !h.HasPipelineRunFinished(a.pipelineRun) || controllerutil.ContainsFinalizer(a.pipelineRun, finalizer) {
+		return controller.ContinueProcessing()
+	}
+
+	err := retry.OnError(retry.DefaultRetry, func(_ error) bool { return true }, func() error {
+		patch := client.MergeFrom(a.pipelineRun.DeepCopy())
+		if !controllerutil.AddFinalizer(a.pipelineRun, finalizer) {
```

Comment:
The `controllerutil.AddFinalizer` function returns a boolean indicating if the finalizer was added or if it was already there ([docs](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#AddFinalizer)), so we shouldn't treat  `false` as an error here, but as an indication if we want to patch the resource on the cluster or not.

Reply:
From the docs:
>  It returns an indication of whether it updated the object's list of finalizers. 
Since this function should only run right after the resource is created, I think it's safe to assume `false` means "we couldn't add the finalizer" rather than "we didn't add the finalizer because it was already there."  Do you think the latter could occur?

Reply:
It could occur, because we cannot expect the predicates to prevent from reconciling the resource more than one time after creation. 
E.g. if the service restarts, it will try to reconcile all existing CRs (disregarding the predicates), so this could end up trying to add the finalizer after it was already added.

Reply:
Since it's not possible to differentiate between the `controllerutil.AddFinalizer` call failing to add the finalizer and it not being added because it's already present on the resource, I propose adopting the approach similar to [this one](https://github.com/redhat-appstudio/release-service/blob/9a0f08573e4ca3f0d5deda2cbda2575b7e7093bd/controllers/release/adapter.go#L131-L150) where we would check if the `a.pipelineRun` doesn't already have the finalizer set on it before doing anything further.